cli/src/commands/bookmark/mod.rs

```diff
@@ -201,3 +203,92 @@ fn is_fast_forward(repo: &dyn Repo, old_target: &RefTarget, new_target_id: &Comm
         true
     }
 }
+
+fn validate_bookmark_names(
+    command: &CommandHelper,
+    ui: &mut Ui,
+    bookmark_names: &[String],
+    command_name: &str,
+) -> Result<(), CommandError> {
+    let helper = command.workspace_helper_no_snapshot(ui)?;
+
+    let mut remote_like = Vec::new();
+    let mut revset_like = Vec::new();
+
+    for name in bookmark_names {
+        match BookMarkKind::parse(name, &helper)? {
+            BookMarkKind::RemoteLike => remote_like.push(name.as_str()),
+            BookMarkKind::RevsetLike => revset_like.push(name.as_str()),
+            BookMarkKind::Normal => {}
+        }
+    }
+
+    if !remote_like.is_empty() {
+        let mut writer = ui.warning_with_heading("Bookmarks containing '@':");
+        for name in &remote_like {
+            writeln!(writer, "  {name}")?;
+        }
+        drop(writer);
+
+        let mut hint = ui.hint_default();
+        writeln!(
+            hint,
+            "To track remote bookmarks, use: jj bookmark --track {}",
+            remote_like.join(" ")
+        )?;
+    }
+
+    if !revset_like.is_empty() {
+        let mut writer = ui.warning_with_heading("Bookmarks resembling revsets:");
+        for name in &revset_like {
+            writeln!(writer, "  {name}")?;
+        }
+        drop(writer);
+
+        let mut warn = ui.warning_default();
+        writeln!(
+            warn,
+            "These names look like revset expressions and might cause confusion: {}",
+            revset_like.join(" ")
+        )?;
+        let mut hint = ui.hint_default();
+
+        if revset_like.len() == 1 {
+            writeln!(
+                hint,
+                "Maybe you meant to use `jj bookmark {command_name} -r {name}` instead?",
+                name = revset_like[0],
+            )?;
+        }
+    }
+
+    Ok(())
+}
+
+enum BookMarkKind {
+    RevsetLike,
+    RemoteLike,
+    Normal,
+}
+
+impl BookMarkKind {
+    fn parse(name: &str, helper: &WorkspaceCommandHelper) -> Result<Self, CommandError> {
+        let node = helper.parse_revset_expression(name)?;
```

Comment:
Perhaps, we can add public `revset::parse_symbol() -> Result<String, ..>` function that does `parse_program()` and `expect_literal()` (or reexport these functions.) Revset aliases shouldn't be expanded, and the other context parameters are unneeded.

Maybe we can add error hint by inspecting the source string? It might be easier if parsed tree (= result of `parse_program()`) is available, but I'm not sure.